cmd/serverutil/listen.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serverutil
+
+import (
+	"context"
+	"net"
+
+	"github.com/golang/glog"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// Listen binds to listenAddr and returns a gRPC connection to it.
+func Listen(ctx context.Context, listenAddr, certFile string) (net.Listener, *grpc.ClientConn, func()) {
+	// Listen and create empty grpc client connection.
+	lis, err := net.Listen("tcp", listenAddr)
+	if err != nil {
+		glog.Exitf("error creating TCP listener: %v", err)
+	}
+	addr := lis.Addr().String()
+	glog.Infof("Listening on %v", addr)
+	// Non-blocking dial before we start the server.
+	tcreds, err := credentials.NewClientTLSFromFile(certFile, "localhost")
+	if err != nil {
+		glog.Exitf("Failed opening cert file %v: %v", certFile, err)
+	}
+	conn, err := grpc.DialContext(ctx, addr, grpc.WithTransportCredentials(tcreds))
```

Comment:
It seems surprising to me that a func called "Listen()" also dials the server. It also does so before a gRPC server is even listening on the socket. It seems that this func would be cleaner if it were split in two.

Reply:
I thought about this too. The reason these are grouped together is that the credentials and addressing for the two are very linked and were nice to keep together.

Reply:
Mention it in the function comment? Or call it DialAndListen()? (I like this less, but it's a common pattern in some codebases I've worked on.)